cli/lsp/config.rs

```diff
@@ -139,25 +145,178 @@ impl WorkspaceSettings {
   }
 }
 
+#[derive(Debug, Clone, Default)]
+pub struct ConfigSnapshot {
+  pub client_capabilities: ClientCapabilities,
+  pub root_uri: Option<Url>,
+  pub settings: Settings,
+}
+
+impl ConfigSnapshot {
+  pub fn specifier_enabled(&self, specifier: &ModuleSpecifier) -> bool {
+    if let Some(settings) = self.settings.specifiers.get(specifier) {
+      settings.1.enable
+    } else {
+      self.settings.workspace.enable
+    }
+  }
+}
+
+enum ConfigRequest {
+  Specifier(ModuleSpecifier, ModuleSpecifier),
+  Workspace,
+}
+
 #[derive(Debug, Default, Clone)]
+pub struct Settings {
+  pub specifiers:
+    HashMap<ModuleSpecifier, (ModuleSpecifier, SpecifierSettings)>,
+  pub workspace: WorkspaceSettings,
+}
+
+#[derive(Debug)]
 pub struct Config {
   pub client_capabilities: ClientCapabilities,
   pub root_uri: Option<Url>,
-  pub specifier_settings: HashMap<ModuleSpecifier, SpecifierSettings>,
-  pub workspace_settings: WorkspaceSettings,
+  settings: Arc<Mutex<Settings>>,
+  tx: mpsc::UnboundedSender<ConfigRequest>,
 }
 
 impl Config {
-  #[allow(unused)]
-  pub fn contains(&self, specifier: &ModuleSpecifier) -> bool {
-    self.specifier_settings.contains_key(specifier)
+  pub fn new(client: lspower::Client) -> Self {
+    let (tx, mut rx) = mpsc::unbounded_channel::<ConfigRequest>();
+    let settings = Arc::new(Mutex::new(Settings::default()));
+    let settings_ref = settings.clone();
+
+    let _join_handle = thread::spawn(move || {
+      let runtime = create_basic_runtime();
+
+      runtime.block_on(async {
+        loop {
+          match rx.recv().await {
+            None => break,
+            Some(ConfigRequest::Workspace) => {
+              let mut settings = settings_ref.lock().unwrap();
+              let specifiers: Vec<(
+                ModuleSpecifier,
+                (ModuleSpecifier, SpecifierSettings),
+              )> = settings.specifiers.clone().into_iter().collect();
+              let mut items = vec![lsp::ConfigurationItem {
+                scope_uri: None,
+                section: Some(SETTINGS_SECTION.to_string()),
+              }];
+              for (_, (uri, _)) in specifiers.iter() {
+                items.push(lsp::ConfigurationItem {
+                  scope_uri: Some(uri.clone()),
+                  section: Some(SETTINGS_SECTION.to_string()),
+                });
+              }
+              let specifiers: Vec<(ModuleSpecifier, ModuleSpecifier)> =
+                specifiers.into_iter().map(|(s, (u, _))| (s, u)).collect();
+              if let Ok(configs) = client.configuration(items).await {
```

Comment:
Awaits while holding the `settings` mutex. I'd try to avoid that at all costs because it's too easy to (re)introduce deadlocks that way. Would releasing and reacquiring the mutex work?

As a general observation: the critical section is pretty long. Shorter is better.

Reply:
I had thought about that, and yeah, you are probably right that it needs to be re-worked to re-acquire the lock. I wrote the code to deal with re-acquiring the lock (e.g. there is a risk something might have changed), but I didn't release it and now that doesn't really make any sense.

Shorter is always better, it expanded trying to deal with the possibility that things could change between when the request was formed and when the response was processed, which added quite a few lines, but I will go over it again and see what can be done to deal with it.  The problem is that hashmaps are unordered which added a level of complexity, but maybe a btree would actually be more useful here, so the keys always come back in a predictable order.